Dashboards: Unify angular auto-migration code

ryantxu opened the pull request:
**What is this feature?**

When angular is disabled, we should automatically migrate some old panels to their new/supported flavors.

**Why do we need this feature?**

We want disabling angular to be as easy as possible


NOTE: this does not verify parity between old+new options, but at least shows something and will trigger the migrations within each panel.

![2023-03-23_14-06-44 (1)](https://user-images.githubusercontent.com/705951/227361613-52925d7b-3ed0-4b2c-bc11-0f71a9ae7222.gif)

ryantxu committed:
auto migrate

ryantxu committed:
auto migrate

ryantxu committed:
auto migrate

tolzhabayev commented:
To the EM's / teams owning this panel plugins - the only ask I have is please inform CX about what will happen to customers and their dashboards that will be auto-migrated, maybe try it out yourself first to have a list of cases that work well and some that will be tricky / need some manual work.

nmarrs commented on public/app/features/dashboard/state/PanelModel.ts:
```diff
@@ -134,6 +134,14 @@ const defaults: any = {
 };
 
+const autoMigrateAngular: Record<string, string> = {
+  graph: 'timeseries',
```
maybe? not sure if change is necessary 
```suggestion
  'graph': 'timeseries',
```

ryantxu replied:
🤣 seems fine, but will immediately get changed back by prettier

nmarrs commented on public/app/features/dashboard/state/PanelModel.ts:
```diff
@@ -431,7 +447,7 @@ export class PanelModel implements DataConfigSource, IPanelModel {
     if (this.autoMigrateFrom) {
-      const wasAngular = this.autoMigrateFrom === 'graph' || this.autoMigrateFrom === 'table-old';
+      const wasAngular = autoMigrateAngular[this.autoMigrateFrom] != null;
       this.callPanelTypeChangeHandler(
```
Have we ever auto migrated angular panels to another panel type in the past? If so it may be useful to look at the approach we took and learn from it for this time around

ryantxu replied:
graph + table-old have been loading this way for a few years, heatmap for 6+ months

The difference here making the pattern more general and adding worldmap+piechart explicitly

ryantxu committed:
Merge remote-tracking branch 'origin/main' into auto-migrate-piechart-worldmap

torkelo committed:
Merge branch 'main' of github.com:grafana/grafana into auto-migrate-piechart-worldmap

torkelo committed:
fix pie chart migration legend not showing

torkelo reviewed (commented):
Took this for a spin and it works pretty great, if I uninstall the old pie chart panel plugin they are automatically migrated to core pie chart. Discovered a small bug in the migration, the legend was not showing for legends that had no value or percentage display. 

The test case for this seems to be in-accurate, I was able to build a pie chart with legend without values (in old pie chart) 

![image](https://user-images.githubusercontent.com/10999/226554240-d6e5212c-534c-4065-ba5c-932f2802ce88.png)

torkelo reviewed (commented):
Wonder if we should rename autoMigrateGraphPanels to autoMigrateOldPanels (and have it both cover graph and table-old, or create a new feature toggle for forced table migration).

ryantxu commented:
> Wonder if we should rename 

YES!  https://github.com/grafana/grafana/pull/65243

ryantxu commented:
See also https://github.com/grafana/grafana/pull/65245

ryantxu committed:
update test

ryantxu committed:
update test

ryantxu committed:
Merge remote-tracking branch 'origin/main' into auto-migrate-piechart-worldmap

ryantxu committed:
update test

ryantxu committed:
Merge remote-tracking branch 'origin/main' into auto-migrate-piechart-worldmap

ryantxu commented:
@nmarrs @torkelo -- I marked this as ready for review since it only improves what we have.  We still need to test and improve the actual migrations from each source, but showing something seems better than breaking

nmarrs reviewed (approved):
Code changes look good to me

ryantxu committed:
Merge remote-tracking branch 'origin/main' into auto-migrate-piechart-worldmap

ryantxu committed:
migrate in dashboard loader

ryantxu committed:
remove duplicate processing

ryantxu committed:
make it an object

ryantxu commented:
@torkelo -- i started poking at the tests, and this evolved enough it would be great to take another look :)

Specifically, I:
* moved the logic from PanelModel to DashboardModel so it can run *after* dashboard migrator has made other changes
* added an option to the DashbodModel constructor so that scenes can force this rather than having custom handling

ryantxu commented on public/app/features/scenes/dashboard/DashboardsLoader.ts:
```diff
@@ -262,14 +263,6 @@ export function createVizPanelFromPanelModel(panel: PanelModel) {
     maxDataPoints: panel.maxDataPoints ?? undefined,
   });
 
-  // Migrate graph to timeseries
```
we can avoid this... and force the other panels by adding the new option above

ryantxu committed:
fix meta

ryantxu committed:
another test

torkelo commented:
@ryantxu looks good, will try to test the migrations themselves more next week and fix issues. I think many of the migrations leave all the angular options intact in the JSON model for example :)

leeoniya commented:
> I think many of the migrations leave all the angular options intact in the JSON model for example :)

we need a SchemaScrubber :D

ryantxu committed:
Merge remote-tracking branch 'origin/main' into auto-migrate-piechart-worldmap
